JHipster 4.0.7 with Gradle+Kafka fails with "object is not an instance of declaring class"

##### **Overview of the issue**

```
[  restartedMain] c.l.m.gateway.config.WebConfigurer       : Web application fully configured
[  restartedMain] c.l.m.g.c.c.CassandraConfiguration       : Configuring Cassandra session
[  restartedMain] c.l.m.gateway.config.CacheConfiguration  : Starting HazelcastCacheManager
[  restartedMain] i.g.j.c.apidoc.SwaggerConfiguration      : Starting Swagger
[  restartedMain] i.g.j.c.apidoc.SwaggerConfiguration      : Started Swagger in 5 ms
[  restartedMain] c.n.c.sources.URLConfigurationSource     : No URLs will be polled as dynamic configuration sources.
[  restartedMain] c.l.gateway.GatewayApp      : The following profiles are active: swagger,dev
[  restartedMain] c.l.gateway.GatewayApp      : Started GatewayApp in 0.55 seconds (JVM running for 16.427)
[  restartedMain] ationConfigEmbeddedWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.context.ApplicationContextException: Failed to start bean 'outputBindingLifecycle'; nested exception is java.lang.IllegalArgumentException: object is not an instance of declaring class
[  restartedMain] c.n.c.sources.URLConfigurationSource     : No URLs will be polled as dynamic configuration sources.
[  restartedMain] com.netflix.discovery.DiscoveryClient    : Saw local status change event StatusChangeEvent [timestamp=1488532023212, current=DOWN, previous=STARTING]
[  restartedMain] com.netflix.discovery.DiscoveryClient    : Shutting down DiscoveryClient ...
[  restartedMain] com.netflix.discovery.DiscoveryClient    : Unregistering ...
[nfoReplicator-0] com.netflix.discovery.DiscoveryClient    : DiscoveryClient_GATEWAY/gateway:9d7acdc18183881572cd3ff78020ea56: registering service...
[  restartedMain] com.netflix.discovery.DiscoveryClient    : DiscoveryClient_GATEWAY/gateway:9d7acdc18183881572cd3ff78020ea56 - deregister  status: 404
[  restartedMain] com.netflix.discovery.DiscoveryClient    : Completed shut down of DiscoveryClient
[nfoReplicator-0] com.netflix.discovery.DiscoveryClient    : DiscoveryClient_GATEWAY/gateway:9d7acdc18183881572cd3ff78020ea56 - registration status: 204
[  restartedMain] com.hazelcast.core.LifecycleService      : [127.0.0.1]:13781 [dev] [3.7.5] [127.0.0.1]:13781 is SHUTTING_DOWN
[  restartedMain] com.hazelcast.instance.Node              : [127.0.0.1]:13781 [dev] [3.7.5] Shutting down connection manager...
[  restartedMain] com.hazelcast.instance.Node              : [127.0.0.1]:13781 [dev] [3.7.5] Shutting down node engine...
[  restartedMain] com.hazelcast.instance.NodeExtension     : [127.0.0.1]:13781 [dev] [3.7.5] Destroying node NodeExtension.
[  restartedMain] com.hazelcast.instance.Node              : [127.0.0.1]:13781 [dev] [3.7.5] Hazelcast Shutdown is completed in 51 ms.
[  restartedMain] com.hazelcast.core.LifecycleService      : [127.0.0.1]:13781 [dev] [3.7.5] [127.0.0.1]:13781 is SHUTDOWN
[  restartedMain] c.l.m.gateway.config.CacheConfiguration  : Closing Cache Manager
[  restartedMain] o.s.boot.SpringApplication               : Application startup failed

org.springframework.context.ApplicationContextException: Failed to start bean 'outputBindingLifecycle'; nested exception is java.lang.IllegalArgumentException: object is not an instance of declaring class
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:176)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:51)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:346)
        at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:149)
        at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:112)
        at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:879)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.finishRefresh(EmbeddedWebApplicationContext.java:144)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:545)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:737)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:370)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:314)
        at com.gateway.GatewayApp.main(GatewayApp.java:74)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
Caused by: java.lang.IllegalArgumentException: object is not an instance of declaring class
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.cloud.stream.binder.kafka.admin.Kafka10AdminUtilsOperation.errorCodeFromTopicMetadata(Kafka10AdminUtilsOperation.java:82)
        at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createTopicAndPartitions(KafkaMessageChannelBinder.java:443)
        at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createTopicsIfAutoCreateEnabledAndAdminUtilsPresent(KafkaMessageChannelBinder.java:421)
        at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createProducerDestinationIfNecessary(KafkaMessageChannelBinder.java:229)
        at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createProducerDestinationIfNecessary(KafkaMessageChannelBinder.java:93)
        at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindProducer(AbstractMessageChannelBinder.java:101)
        at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindProducer(AbstractMessageChannelBinder.java:55)
        at org.springframework.cloud.stream.binder.AbstractBinder.bindProducer(AbstractBinder.java:150)
        at org.springframework.cloud.stream.binding.BindingService.bindProducer(BindingService.java:125)
        at org.springframework.cloud.stream.binding.BindableProxyFactory.bindOutputs(BindableProxyFactory.java:225)
        at org.springframework.cloud.stream.binding.OutputBindingLifecycle.start(OutputBindingLifecycle.java:57)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:173)
        ... 17 common frames omitted
```

##### **Motivation for or Use Case**

My gateway and microservice would not start, be it standalone or in Docker. Both setups throw the same error. I am using Cassandra, Kafka, and JHipster Registry. My standalone setup (run via `./gradlew`) accesses services created via Docker Compose.

As a side test I also created a new gateway from scratch and ran the whole thing in Docker Compose and it too threw the same error.

##### **Reproduce the error**

Upgrade to 4.0.7 or create one from scratch. I am only using microservices.

##### **Related issues**

I am not sure...

##### **Suggest a Fix**

The erring line is this: `Environment env = app.run(args).getEnvironment();`.

It might be related to a configuration property being set to a wrong type but I can't tell which is which.

##### **JHipster Version(s)**

4.0.7

##### **JHipster configuration**

```
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

gateway@0.0.0 /home/...
└── generator-jhipster@4.0.7 

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

yaml
{
  "generator-jhipster": {
    "jhipsterVersion": "4.0.7",
    "baseName": "gateway",
    "packageName": "com.gateway",
    "packageFolder": "com/gateway",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "no",
    "clusteredHttpSession": "hazelcast",
    "websocket": "spring-websocket",
    "databaseType": "cassandra",
    "devDatabaseType": "cassandra",
    "prodDatabaseType": "cassandra",
    "searchEngine": false,
    "messageBroker": "kafka",
    "serviceDiscoveryType": "eureka",
    "buildTool": "gradle",
    "enableSocialSignIn": false,
    "jwtSecretKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "useSass": true,
    "clientPackageManager": "yarn",
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling",
      "cucumber",
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "fr",
      "tr"
    ],
    "clientFramework": "angular2"
  }
}

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

ls: no such file or directory: .jhipster/*.json

##### **Browsers and Operating System**

java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

git version 2.9.3

node: v6.10.0

npm: 3.10.10

yeoman: 1.8.5

yarn: 0.21.3

Docker version 17.03.0-ce, build 60ccb22

docker-compose version 1.10.1, build b252738
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

None.

##### **Browsers and Operating System**

Firefox 51.0.1 (64-bit), Ubuntu 16.10 64-bit

- [x ] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/server/templates/_build.gradle
+++ b/generators/server/templates/_build.gradle
@@ -272,6 +272,7 @@ dependencies {
     <%_ if (messageBroker === 'kafka') { _%>
     compile "org.springframework.cloud:spring-cloud-stream"
     compile "org.springframework.cloud:spring-cloud-stream-binder-kafka"
+    compile "org.springframework.kafka:spring-kafka:${spring_kafka_version}"
     <%_ } _%>
     compile "org.springframework.boot:spring-boot-starter-security"
     compile ("org.springframework.boot:spring-boot-starter-web") {

--- a/generators/server/templates/_gradle.properties
+++ b/generators/server/templates/_gradle.properties
@@ -56,6 +56,7 @@ spring_cloud_version=Camden.SR6
 <%_ } _%>
 <%_ if (messageBroker === 'kafka') { _%>
 spring_cloud_stream_version=Brooklyn.SR2
+spring_kafka_version=1.0.5.RELEASE
 <%_ } _%>
 <%_ if (devDatabaseType === 'mssql' || prodDatabaseType === 'mssql') { _%>
 mssql_jdbc_version=6.1.0.jre8

